torchdata/stateful_dataloader/incremental_state.py

```diff
@@ -0,0 +1,178 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+from typing import Any, Dict, Optional
+
+import torch
+
+_WORKER_ID = "worker_id"
+_FETCHER_STATE = "fetcher_state"
+_FETCHER_ENDED = "fetcher_ended"
+_DATASET_STATE = "dataset_state"
+_DATASET_ITER_STATE = "dataset_iter_state"
+
+
+class Flattener:
+    @classmethod
+    def flatten(cls, nested_data: Optional[Dict[str, Any]], prefix: str = "", separator: str = "/"):
+        if nested_data is None:
+            return None
+        data = {}
+        if isinstance(nested_data, dict):
+            for key, value in nested_data.items():
+                flat = cls.flatten(
+                    value,
+                    prefix=f"{prefix}{separator}{key}" if len(prefix) > 0 else key,
+                    separator=separator,
```

Comment:
Would this break if users have keys with `/`?

Reply:
Another way to do this that might be more robust is to pass this in as expanding tuples, which are also immutable:

```
prefix: Tuple[...]
flatten(value, prefix=prefix + (key,)
```

Reply:
There's also an assumption here that keys are a) strings, b) immutable

Reply:
(responding to myself) I think if it's usable as a key in a dictionary, it should be immutable, see: https://docs.python.org/3/reference/datamodel.html#object.__hash__